proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,671 @@
+polymorphic functions may have variable multiplicity, *e.g.*
+
+::
+
+  map :: (a:p -> b) -> [a]:p -> [b]
+
+without polymorphism we would need two implementations of `map`. With
+the exact same code: one for ``p=1`` and one for ``p=ω``. Function
+composition is even worse: it takes two multiplicity parameters, hence,
+would require four identical implementations:
+
+::
+
+  (.) :: (b:p -> c) -> (a:q -> b) -> a:(p ':* q) -> c
+
+Syntax
+~~~~~~
+
+The new primary constructs are: multiplicities and the multiplicity
+indexed arrow.
+
+- Multiplicities are a datatype:
+
+  ::
+
+    data Multiplicity
+      = One
+      | Omega
+      | Multiplicity :* Multiplicity
+      | Multiplicity :+ Multiplicity
+
+  In the following, for conciseness, we write ``1`` for ``One`` and
+  ``U`` (ASCII) or ``ω`` (Unicode) for ``Omega``. Note: unification
+  of multiplicities will be performed up to the semiring laws.
+- The multiplicity annotated arrow, for polymorphism, is written ``a:p
+  -> b`` (where ``a`` and ``b`` are types and ``p`` is a
+  multiplicity). To avoid introducing a new notion of "mixfix"
+  operators, we introduce a familly of (infix) type constructors:
+  ``(:p ->)`` for each multiplicity ``p``.
+
+The linear and unrestricted arrows are aliases:
+
+- ``(->)`` is an alias for ``(:'U ->)``
+- ``(->.)`` (ASCII syntax) and ``(⊸)`` (Unicode syntax) are aliases
+  for ``(:'1 ->)``
+
+Constructors & pattern-matching
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Constructors of data types defined with the Haskell'98 syntax
+
+::
+
+  data Foo
+    = Bar A B
+    | Baz C
+
+have linear function types, that is ``Bar :: A ->. B ->. Foo``. This
+is true in every module, including those without ``-XLinearTypes``
+turned on. This implies that most types in ``base`` (``Maybe``,
+``[]``, etc…) have linear constructors. We also make primivitive
+tuples ``(,)`` have linear constructors.
+
+With the GADT syntax, multiplicity of the arrows is honored:
+
+::
+
+  data Foo2 where
+    Bar2 :: A ->. B -> C
+
+then ``Bar2 :: A ->. B -> C``
+
+The definition of consuming a value in a data type exactly once must
+be refined to take the multiplicities of fields into account:
+
+- Consuming a value in a datatype exactly once means evaluating it to
+  head normal form and consuming its *linear* fields exactly once
+
+When pattern macthing a linear argument, linear fields are introduced
+as linear variables, and unrestricted fields as unrestricted
+variables:
+
+::
+
+  f :: Foo2 ->. A
+  f (Bar2 x y) = x  -- y is unrestricted, hence does not need to be consumed
+
+Base
+~~~~
+
+Because linear functions only strengthen the contract of unrestricted
+functions, a number of functions of ``base`` can get a more precise
+type. However, for pedagogical reason, to prevent linear types from
+interfering with newcomers' understanding the ``Prelude``, this
+proposal does not modify ``base``. Instead we will release a library
+exposing the stronger types for ``base`` functions.
```

Comment:
Aren't we going to end up with two ecosystems? One for the normal Prelude and one for the strengthened Prelude?

Reply:
I'm really not too worried about that. Because we are not redefining any type: we are just strengthening the type of functions. So both "preludes" will be completely compatible.

We're adding types too, but it must be understood that if any of them make it into `base`, then we will re-export this one instead of defining our own version.

Reply:
Right. But worth a comment in the proposal itself.

Reply:
Done.